### PR TITLE
Add support for ADC on STM32 H7 series

### PR DESCRIPTION
**Pull request by gmarull**
- Add support for ADC on H7 series. Note that ADC1 and ADC2 share the same register set, so it is added as "adc1_2"
- Enables ADC on Nucleo H743ZI board (ADC12 channel 15).

Performed quick tests on Nucleo H743ZI board on ADC12 CH15 and ADC3 CH6 using a signal generator as an analog test signal.

NOTE: I think that current STM32 ADC driver needs a rework, as it's in a kind of "ifdef hell" state.

### CODE REVIEW

**Comment by erwango** [comment]
^^@ABOSTM

File: dts/arm/st/h7/stm32h7.dtsi [hunk 1]
```diff
@@ -546,6 +546,26 @@
 				#pwm-cells = <2>;
 			};
 		};
+
+		adc12: adc@40022000 {
```
Comment (erwango) [inline dts/arm/st/h7/stm32h7.dtsi:550]: I know this is used in STM32 packages, but I wonder if adc1_2 would best for disambiguation.

**Review by erwango — CHANGES_REQUESTED** [review]
Regarding the ADC1/2 common node, impact I have in mind in having one node for both instances is that you can't invoke them separately in an application. Looking to reference manual, it seems though that ADC1 and ADC2 doesn't share all resources and ADC1 could be used with ADC2 in a master/slave mode. Running these instances in parallel is not possible today due to driver limitation, so you're not breaking anything. Also I'm not enough familiar with ADC to say if this mode is really something that is required in a near future. 

So this is fine for now, but maybe we'll have to come back on this later.

Anyway, please use ADC1_2 for disambiguation.

**Comment by gmarull** [comment]
@erwango I'll adjust node label to `ADC1_2`. Regarding `adc12` vs. `adc1_2` I'm fine with both approaches, so just let me know. I think it will be difficult to add support for ADC1/2 Master/Slave mode in a general purpose driver, I think it is designed for some specific applications (e.g. motor control where you want to sample phases in synchronization). Some details: https://www.st.com/resource/en/application_note/cd00258017-stm32s-adc-modes-and-their-applications-stmicroelectronics.pdf

**Review by erwango — CHANGES_REQUESTED** [review]
Let's go with adc1_2

**Review by erwango — APPROVED** [review]

**Comment by gmarull** [comment]
@erwango updates number of channels in the driver to 20, as H7 has 20 valid external channels (F7 or F4 have 19, so the number was actually wrong).

**Comment by stephanosio** [comment]
> -:70: WARNING:SPDX_LICENSE_TAG: Misplaced SPDX-License-Identifier tag - use line 1 instead
> #70: FILE: drivers/adc/adc_stm32.c:7:
>   * SPDX-License-Identifier: Apache-2.0

Please ignore these warnings (see #24921)

**Comment by galak** [comment]
can you rebase to resolve merge conflict.

**Comment by gmarull** [comment]
@galak done

**Pull request merged**